src/EditorFeatures/Core.Wpf/InlineRename/UI/SmartRename/SmartRenameViewModel.cs

```diff
@@ -0,0 +1,134 @@
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.ObjectModel;
+using System.ComponentModel;
+using System.Threading;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.VisualStudio.Text.Editor.SmartRename;
+
+namespace Microsoft.CodeAnalysis.InlineRename.UI.SmartRename;
+
+internal sealed class SmartRenameViewModel : INotifyPropertyChanged, IDisposable
+{
+    public static string GeneratingSuggestions => EditorFeaturesWpfResources.Generating_suggestions;
+
+#pragma warning disable CS0618 // Editor team use Obsolete attribute to mark potential changing API
+    private readonly ISmartRenameSession _smartRenameSession;
+#pragma warning restore CS0618 
+
+    private readonly IThreadingContext _threadingContext;
+
+    private readonly CancellationTokenSource _cancellationTokenSource = new();
+
+    public event PropertyChangedEventHandler? PropertyChanged;
+
+    public event EventHandler<string?>? OnSelectedSuggestedNameChanged;
+
+    public ObservableCollection<string> SuggestedNames { get; } = new ObservableCollection<string>();
+
+    public bool IsAvailable => _smartRenameSession?.IsAvailable ?? false;
+
+    public bool HasSuggestions => _smartRenameSession?.HasSuggestions ?? false;
+
+    public bool IsInProgress => _smartRenameSession?.IsInProgress ?? false;
+
+    public string StatusMessage => _smartRenameSession?.StatusMessage ?? string.Empty;
+
+    public bool StatusMessageVisibility => _smartRenameSession?.StatusMessageVisibility ?? false;
+
+    private string? _selectedSuggestedName;
+
+    /// <summary>
+    /// The last selected name when user click one of the suggestions. <see langword="null"/> if user hasn't clicked any suggestions.
+    /// </summary>
+    public string? SelectedSuggestedName
+    {
+        get => _selectedSuggestedName;
+        set
+        {
+            if (_selectedSuggestedName != value)
+            {
+                _threadingContext.ThrowIfNotOnUIThread();
+                _selectedSuggestedName = value;
+                OnSelectedSuggestedNameChanged?.Invoke(this, value);
+            }
+        }
+    }
+
+    public SmartRenameViewModel(
+        IThreadingContext threadingContext,
+        IAsynchronousOperationListenerProvider listenerProvider,
+#pragma warning disable CS0618 // Editor team use Obsolete attribute to mark potential changing API
+        ISmartRenameSession smartRenameSession)
+#pragma warning restore CS0618
+    {
+        _threadingContext = threadingContext;
+        _smartRenameSession = smartRenameSession;
+        _smartRenameSession.PropertyChanged += SessionPropertyChanged;
+        var listener = listenerProvider.GetListener(FeatureAttribute.SmartRename);
+
+        using var listenerToken = listener.BeginAsyncOperation(nameof(_smartRenameSession.GetSuggestionsAsync));
```

Comment:
Wouldn't this dispose before the async operation completes?

Reply:
AH yeah I forget to call CompletesAsyncOperation